### ISSUE TITLE
Regression in applying `google_sql_database_instance` update without refresh

### ISSUE THREAD

**Issue by AaronFriel** [hashicorp/terraform-provider-google#15762]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave _+1_ or _me too_ comments, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.
* If an issue is assigned to the `modular-magician` user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If an issue is assigned to a user, that user is claiming responsibility for the issue. If an issue is assigned to `hashibot`, a community member has claimed the issue already.

### Terraform Version

- n/a, issue is in provider code

### Affected Resource(s)

* google_sql_database_instance

### Terraform Configuration Files

```tf
resource "google_sql_database_instance" "instance" {
  region              = "us-central1"
  database_version    = "MYSQL_5_7"
  deletion_protection = false
  settings {
    tier = "db-f1-micro"
  }
}
```

Or any trivial configuration. The bug occurs when applying an update after the upstream cloud resource has drifted, see steps to reproduce.

### Debug Output

```
google_sql_database_instance.instance: Modifying... [id=bug-412-2023-09-07]
╷
│ Error: Error, failed to update instance settings for : googleapi: Error 412: Condition does not match., staleData
│
│   with google_sql_database_instance.instance,
│   on main.tf line 10, in resource "google_sql_database_instance" "instance":
│   10: resource "google_sql_database_instance" "instance" {
│
╵
```

### Expected Behavior

In both the google provider and the google-beta provider, a read(s) to acquire the cloud resource's `settings.version` is conditional. However, prior to #7200 being merged, this was not the case. The change from `GetOk()` to `HasChange()` resulted in what would have been a virtually unconditional read and then patch with the correct settings version into one that only occurs when one of several properties change.

### Before

In `resourceSqlDatabaseInstanceUpdate`, a read occurs if `database_version` is present:

https://github.com/hashicorp/terraform-provider-google/blob/90c49f7aaadec2c8cf7a5a2157f3fd0977e6a1db/google/resource_sql_database_instance.go#L1511-L1513

https://github.com/hashicorp/terraform-provider-google/blob/90c49f7aaadec2c8cf7a5a2157f3fd0977e6a1db/google/resource_sql_database_instance.go#L1526-L1530

https://github.com/hashicorp/terraform-provider-google/blob/90c49f7aaadec2c8cf7a5a2157f3fd0977e6a1db/google/resource_sql_database_instance.go#L1614-L1622

After:
https://github.com/hashicorp/terraform-provider-google/blob/ab21e8a307dfd90cdd12cd05d9a40a6bc18a3af1/google/resource_sql_database_instance.go#L1574-L1576

### After

A read occurs only if the field has changed:

https://github.com/hashicorp/terraform-provider-google/blob/90c49f7aaadec2c8cf7a5a2157f3fd0977e6a1db/google/resource_sql_database_instance.go#L1589-L1592

(The next two blocks are unchanged, the meaningful diff is the condition on 1576.)

### Actual Behavior

On applying updates without a refresh, due to exigent circumstances (external modification, drift, other), the update fails with the error above.

### Steps to Reproduce

Here is a a repository and a shell script to reproduce this issue:

https://github.com/AaronFriel/google-cloud-sql-instance-412-bug

The script uses Terraform.

### Important Factoids

While root causing this issue, I attempted to write an acceptance test and make a pull request with it. Unfortunately there is no clean way to write an acceptance test for this scenario!

When executing a test like so:

```go
resource.test(t, resource.TestCase{
  ...,  Steps: []resource.TestStep{
    { 
      Config: firstConfig
    },
    { 
      PreConfig: func() {
        // modify the upstream resource via API call directly
        // to simulate drift
      },
      Config: secondConfig
    },
    ...
  }
}
```

The Terraform Plugin SDKv2 test harness unconditionally performs a `terraform refresh` prior to plan and apply:

https://github.com/hashicorp/terraform-plugin-sdk/blob/af738e0d482f699504d9e35e134766da459ef1f6/helper/resource/testing_new_config.go#L27-L31

Even when `PreConfig()` is defined, it runs _prior_ to the refresh.

I do think this is a bug, but I leave it to the authors here to file an issue on the Plugin SDK. It's not uncommon for users to [skip automatic state refresh](https://developer.hashicorp.com/terraform/cloud-docs/run/modes-and-options#skipping-automatic-state-refresh), so being unable to test that scenario is a challenge for users.

### References

I discovered the issue from a bug report by users in Pulumi who discovered this regression:
- https://github.com/pulumi/pulumi-gcp/issues/794

**Comment by AaronFriel** [comment]
@melinath That's correct, Pulumi skips refresh by default. It's not an [uncommon configuration (GitHub Code Search)](https://github.com/search?q=%2Fterraform.*-refresh%3Dfalse%2F&type=code) in the Terraform world, for a variety of reasons.

**Comment by AaronFriel** [comment]
@melinath this is affecting Pulumi users, possible to reconsider this?

**Comment by AaronFriel** [comment]
@melinath we patched this in our fork for Pulumi users, possible to reconsider upstreaming the fix?

**Comment by melinath** [comment]
@AaronFriel we'd definitely accept a patch via https://github.com/GoogleCloudPlatform/magic-modules.

**Comment by tobymaro** [comment]
@AaronFriel, can you help with this?

**Comment by deleocj** [comment]
Hello. Are there any updates on resolving this?
